### ISSUE TITLE
Cannot update npm (node 8.4.0)

### ISSUE THREAD

**Issue by t0lkman** [coreybutler/nvm-windows#300]
If this is a question about how to use NVM4W, please use [stackoverflow](http://stackoverflow.com/questions/tagged/nvm-windows) instead.

If this is an issue regarding antivirus, make sure you search the existing issues first.

### My Environment

- [ ] Windows 7 or below (not truly supported due to EOL - see wiki for details)
- [ ] Windows 8
- [ ] Windows 8.1
- [x] Windows 10
- [ ] Windows 10 IoT Core
- [ ] Windows Server 2012
- [ ] Windows Server 2012 R2
- [ ] Windows Server 2016

- [ ] My Windows installation is non-English.

### I'm using NVM4W version:

- [ ] 1.1.6
- [x] 1.1.5
- [ ] 1.1.4
- [ ] 1.1.3
- [ ] 1.1.2
- [ ] 1.1.1
- [ ] Older
- [ ] OTHER (Please Specify)

### I have already...

- [x] read the README to be aware of npm gotchas & antivirus issues.
- [ ] reviewed the [wiki](https://github.com/coreybutler/nvm-windows/wiki) to make sure my issue hasn't already been resolved.
- [x] verified I'm using an account with administrative privileges.
- [x] searched the issues (open and closed) to make sure this isn't a duplicate.
- [ ] made sure this isn't a question about how to use NVM for Windows, since [gitter](https://gitter.im/coreybutler/nvm-windows?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) is used for questions and comments.

### My issue is related to (check only those which apply):

- [ ] settings.txt
- [ ] proxy support (Have you tried version 1.1.0+?)
- [ ] 32 or 64 bit support (Have you tried version 1.1.3+?)
- [ ] Character escaping (Have you tried version 1.1.6+?)
- [x] A standard shell environment (terminal/powershell)
- [ ] A non-standard shell environment (Cmder, Hyper, Cygwin, git)

### Expected Behavior

npm i -g npm@latest should install latest npm version

### Actual Behavior
an error:
`
npm ERR! path C:\Program Files\nodejs\npm.cmd
npm ERR! code EEXIST
npm ERR! Refusing to delete C:\Program Files\nodejs\npm.cmd: is outside C:\Program Files\nodejs\node_modules\npm and not a link
npm ERR! File exists: C:\Program Files\nodejs\npm.cmd
npm ERR! Move it away, and try again.
`

### Steps to reproduce the problem:
run as an administrator `npm i -g npm@latest `

**Comment by manuelbieh** [comment]
Same problem here.

**Comment by kuncevic** [comment]
I had the same issue, and here is how you update (hack)

run `npm install -g npm`
**while it is running** remove `npm.cmd` and `npm` from `c:\Program Files\nodejs\` then remove `.bin`
from `c:\Program Files\nodejs\node_modules\npm\`, then it should succeed

**Comment by alexgorbatchev** [comment]
Seeing same problem

**Comment by cannap** [comment]
same windows 10 node 8.5.0 current npm 5.3.0

**Comment by 19majkel94** [comment]
The easier solution is to copy dir `node_modules/npm` from the nodejs location, remove the npm bin and cmd, and run `node npm-cli.js i -g npm@latest` inside bin in the copied folder.

**Comment by wayofthefuture** [comment]
So much for production. Good luck doing in a script.

**Comment by wayofthefuture** [comment]
They should rename continuous integration to continuous intebreaktion

**Comment by wayofthefuture** [comment]
Solution:

```javascript
cd c:\programdata\nvm\v8.4.0
ren npm.cmd npm-old.cmd
ren npm npm-old
npm-old install -g npm@5.4.2
npm -v
```

Viola, yo.

**Comment by wayofthefuture** [comment]
@19majkel94 @kuncevic I had no luck with the solutions above. Any ideas how to manually update this?

**Comment by wayofthefuture** [comment]
@19majkel94 Couldn't get it working with your method... any other ideas which I could script?

**Comment by coreybutler** [comment]
It appears as though npm has changed how it updates itself... without telling anyone. npx was also introduced with only a minor version change (as opposed to a major change).... which is _semantically_ correct, but still hard to follow along. 

Apparently there is some file sandboxing happening. I spent the better part of today investigating and attempting a fix, but each fix surfaces another issue. Bear with me, I'll get it fixed as soon as I can. I'm the only one working on this at the moment... and I happen to be in the middle of a different product launch (after 2yrs of work), so a PR would gladly be accepted.

**Comment by wayne-oscme** [comment]
I wish I could help, but I have an observation... I think that the issues are related to the shortcut to "`C:/Program Files/nodejs`".  Perhaps if a junction were used instead of a shortcut?  Windows can't tell the difference between a 'real' directory and a junction.  So, it would look like node was really and truly installed in the default directory.  No more 'operating' in a directory outside of the resolved directory.

[https://docs.microsoft.com/en-us/sysinternals/downloads/junction](https://docs.microsoft.com/en-us/sysinternals/downloads/junction)

**Comment by wayne-oscme** [comment]
BTW, using a junction should future proof nvm4w against any future changes to npm's behavior.

**Comment by 19majkel94** [comment]
@wayne-oscme 
Have you read the readme?

> The second option is to use a symlink. This concept requires putting the symlink in the system PATH, then updating its target to the node installation directory you want to use. This is a straightforward approach, and seems to be what people recommend.... until they realize just how much of a pain symlinks are on Windows. This is why it hasn't happened before.

> In order to create/modify a symlink, you must be running as an admin, and you must get around Windows UAC (that annoying prompt). Luckily, this is a challenge I already solved with some helper scripts in node-windows. As a result, NVM for Windows maintains a single symlink that is put in the system PATH during installation only. Switching to different versions of node is a matter of switching the symlink target. As a result, this utility does not require you to run nvm use x.x.x every time you open a console window. When you do run nvm use x.x.x, the active version of node is automatically updated across all open console windows. It also persists between system reboots, so you only need to use nvm when you want to make a change.

**Comment by wayne-oscme** [comment]
@19majkel94 
Please read about junctions.  The 'symlink' aka window's shortcut has problems re resolution of target directory from processes.  While window's explorer and terminal seem to handle them nicely, they aren't handled nicely in other processes.  Open notepad, choose open file to get the standard window's file dialog.  Navigate to `"C:/Program Files/nodejs"`  and see where you end up.

Being a redirect to the target directory, which as we see is where processes end up, can lead to some interesting issues.

Junctions are the equivalent of hard links in *nix.  Windows only supports these to directories, and not to files, but that's all we need here.  They are not redirects.  They are, for all intents and purposes, the directory, no different than the path created when the directory was first created.

They don't redirect, they are the destination.  They can be deleted and created again with each 'use' command.  Therefore, they will have the same effects and capabilities as shortcuts, but none of the downfalls that are creating the current problem.

**Comment by coreybutler** [comment]
This has nothing to do with junctions/hard links/symlinks. @kuncevic provided a screenshot citing npm errors, the key part being `Refusing to delete ... which is outside of ....`. That is a hard-coded message in npm, preventing further operations. Bottom line: npm is refusing to full update itself.

This didn't seem to be an issue until npm 5.x.x, and I'm still digging through the details to figure out what has changed. I may have to write a workaround (i.e. hack), which I'm never fond of. I'm also looking into utilizing the msi packages, which seems to be the only way to get the attention of the folks at npm.

If you're in dire need of updating npm, you can download it via `https://github.com/npm/npm/archive/vX.X.X.zip` and extract it into the `node_modules` directory of your node installation root. You'll need to manually create a symlink in the _node installation root_ to `npm.cmd` and `npx.cmd`. The final file structure should look something like:

```
nvm
- v8.5.0
   - node_modules
       - npm
          - bin
               - npm.cmd
               - npx.cmd
          ...
   - node.exe
   - npm.cmd (symlink to v8.5.0/node_modules/npm/bin/npm.cmd)
   - npx.cmd (symlink to v8.5.0/node_modules/npm/bin/npx.cmd)
```

**Comment by wayne-oscme** [comment]
Okay, I guess I should try before selling.  I replaced the shortcut with a
junction, and got the same exact issue.  Thought I had something there.
Sorry to waste your time.

**Comment by kuncevic** [comment]
Just updated 5.4.1 → 5.4.2 with no issues.

```
c:\>npm install npm -g
C:\Program Files\nodejs\npx -> C:\Program Files\nodejs\node_modules\npm\bin\npx-cli.js
C:\Program Files\nodejs\npm -> C:\Program Files\nodejs\node_modules\npm\bin\npm-cli.js
+ npm@5.4.2
removed 1 package and updated 2 packages in 12.125s
```

**Comment by coreybutler** [comment]
@kuncevic - thanks... that's indicative the problem must be within npm 5.4.3+.

**Comment by wayne-oscme** [comment]
npm -g install npm@5.4.2 still fails for me, same errors.

**Comment by doug2k1** [comment]
For me it fails from Node 8.4.0 and up. It installs NPM 5.3.0, and then won't allow me to upgrade it.
Older Node versions (I have 4.8.4) allow NPM to update normally.

**Comment by manuelbieh** [comment]
Man. npm 5 and nvm really sucks big time. I've downgraded to npm 4.6.2 last week and just tried to give 5.4.2 another try. Looks like npm has been *uninstalled* while trying to install react-native-cli, a completely different package:

```
Manuel@Manuel-406 /cygdrive/d/htdocs/wallet
$ npm install -g react-native-cli
C:\Program Files\nodejs\npx -> C:\Program Files\nodejs\node_modules\npm\bin\npx-cli.js
npm WARN Error: EPERM: operation not permitted, scandir 'C:\Program Files\nodejs\node_modules\npm\node_modules\libnpx\node_modules\yargs\node_modules\os-locale\node_modules\execa\node_modules\cross-spawn\node_modules\shebang-command\node_modules'
npm WARN  { Error: EPERM: operation not permitted, scandir 'C:\Program Files\nodejs\node_modules\npm\node_modules\libnpx\node_modules\yargs\node_modules\os-locale\node_modules\execa\node_modules\cross-spawn\node_modules\shebang-command\node_modules'
npm WARN   stack: 'Error: EPERM: operation not permitted, scandir \'C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\libnpx\\node_modules\\yargs\\node_modules\\os-locale\\node_modules\\execa\\node_modules\\cross-spawn\\node_modules\\shebang-command\\node_modules\'',
npm WARN   errno: -4048,
npm WARN   code: 'EPERM',
npm WARN   syscall: 'scandir',
npm WARN   path: 'C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\libnpx\\node_modules\\yargs\\node_modules\\os-locale\\node_modules\\execa\\node_modules\\cross-spawn\\node_modules\\shebang-command\\node_modules' }
npm ERR! path C:\Program Files\nodejs\npm
npm ERR! code EPERM
npm ERR! errno -4048
npm ERR! syscall open
npm ERR! Error: EPERM: operation not permitted, open 'C:\Program Files\nodejs\npm'
npm ERR!  { Error: EPERM: operation not permitted, open 'C:\Program Files\nodejs\npm'
npm ERR!   cause:
npm ERR!    { Error: EPERM: operation not permitted, open 'C:\Program Files\nodejs\npm'
npm ERR!      errno: -4048,
npm ERR!      code: 'EPERM',
npm ERR!      syscall: 'open',
npm ERR!      path: 'C:\\Program Files\\nodejs\\npm' },
npm ERR!   stack: 'Error: EPERM: operation not permitted, open \'C:\\Program Files\\nodejs\\npm\'',
npm ERR!   errno: -4048,
npm ERR!   code: 'EPERM',
npm ERR!   syscall: 'open',
npm ERR!   path: 'C:\\Program Files\\nodejs\\npm',
npm ERR!   parent: 'v8.5.0' }
npm ERR!
npm ERR! Please try running this command again as root/Administrator.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Manuel\AppData\Roaming\npm-cache\_logs\2017-10-02T14_54_04_833Z-debug.log

Manuel@Manuel-406 /cygdrive/d/htdocs/_git/wallet
$ npm install -g react-native-cli
sh: npm: Kommando nicht gefunden.
```

**Comment by yavorski** [comment]
`ubuntu`: `16.04.2 LTS` `nodejs`: `8.6.0` `npm`: `5.3.0` 

``` shell
npm i -g npm@latest

/home/user/.npm-global/bin/npm -> /home/user/.npm-global/lib/node_modules/npm/bin/npm-cli.js
/home/user/.npm-global/bin/npx -> /home/user/.npm-global/lib/node_modules/npm/bin/npx-cli.js
+ npm@5.4.2
added 68 packages, removed 164 packages and updated 14 packages in 14.307s
```

`npm -v` still prints `5.3.0` and it's not updated?

**Comment by coreybutler** [comment]
@yavorski - this is nvm for Windows. I think you want [nvm for *nix](https://github.com/creationix/nvm).

**Comment by coreybutler** [comment]
I have confirmed this issue is sandboxing within npm 5.x.x. Prior editions of npm did not have this. Unfortunately, the fix is non-trivial, and judging by @yavorski's experience, it's not specific to this project.

**Comment by yavorski** [comment]
Yep no nvm in my case. Linux + Nodejs from nodesource.

**Comment by larryboymi** [comment]
Having same issue here...

**Comment by cdswyda** [comment]
npm version 5.4.2 
node version 8.7.0 
window 10 

I have the same problem。

**Comment by farnetani** [comment]
Windows 10
`mv npm.cmd "c:\Program Files\nodejs\node_modules\npm\" `

and after

`npm i -g npm@latest`

Solved to me!

**Comment by Mohamed3on** [comment]
I have similar behavior with Node v8.9.0. It comes with npm 5.5.1, and when I try to downgrade to npm 4, it gives me the `refusing to delete` error.

**Comment by thegreatco** [comment]
I tried to manually fix this, by pulling npm from the official zip file, and I got this error:
![image](https://user-images.githubusercontent.com/1833767/32419437-c9f080fe-c247-11e7-934f-955bb656585d.png)

I suspect this error is occurring silently and corrupting the install.

**Comment by RichardMisiak** [comment]
Same problem here
Windows 10 
npm 5.3.0
node 6.11.2

**Comment by thegreatco** [comment]
@RichardMisiak did you try moving the `nvm` install to a path closer to `C:\`? Like `C:\nvm`? It cleared up the issue for me.

**Comment by kuncevic** [comment]
With the latest node `9.1` that bringing the latest `npm 5.5.1` by default there is even more "fun" https://github.com/npm/npm/issues/19019

**Comment by AleCaste** [comment]
My solution is NOT very elegant but it works. Here it is.
After running ```npm install npm@5.6 -g``` I got the following error:
```
npm ERR! path d:\node\npm.cmd
npm ERR! code EEXIST
npm ERR! Refusing to delete d:\node\npm.cmd: is outside d:\node\node_modules\npm and not a link
npm ERR! File exists: d:\node\npm.cmd
npm ERR! Move it away, and try again.
...
```

My workaround was the following:
1. ```npm install -g yarn```
2. Restart the computer
3. Run ```yarn global bin``` , and add the folder that is displayed to the PATH env variable of Windows
4. Go to the folder where npm.cmd is located (check this from the error we just got on the ```npm install npm@5.6 -g``` run: ```d:\node\``` in my case) and:
  2.1. Rename ```npm``` to ```npm.original``` (or just delete it)
  2.1. Rename ```npm.cmd``` to ```npm.cmd.original``` (or just delete it)
5. Close the terminal and open a new one
6. ```yarn global add npm@5.6```
7. ```npm install npm@5.6 -g```   (this will install npm using the npm version located on the ```yarn global bin``` folder)
8. ```yarn global remove npm```   (to delete the npm version installed by yarn, and keep only the version installed by npm itself)

**Comment by rolf-schmidiger** [comment]
Just rename the batch BEFORE upgrading:
run cmd
```
cd %ProgramFiles%\nodejs
ren npm.cmd npm2.cmd
ren npm npm2 
npm2 install npm@latest -g
```
delete the files after upgrade:

```
del npm2
del npm2.cmd
```

**Comment by wbt** [comment]
@rolf-schmidiger In my experience I had to do the same renaming with npx and npx.cmd as with npm and npm.cmd.

**Comment by ambroselittle** [comment]
Deliciously simple and effective, @rolf-schmidiger. Thank you.

@coreybutler, any idea on a solution? Just curious. No pressure.

**Comment by Jameskmonger** [comment]
None of the solutions listed here worked for me, and in the end it was quicker to use `nvmw uninstall` to delete the Node installation and then `nvmw install` to reinstall it.

**Comment by thegreatco** [comment]
Isn't `nvmw` a completely different tool?

**Comment by Jameskmonger** [comment]
@thegreatco Oops! You're right. Interesting that the same issue occurs on both, though!

**Comment by stramel** [comment]
Based on @rolf-schmidiger's answer, I created an update script for it... Kinda sucks that this is the only way I could get it to work.

```sh
#!/usr/bin/bash

cd "$PROGRAMFILES"/nodejs
mv npm.cmd npm2.cmd
mv npm npm2
npm2 install -g npm@latest
rm npm2
rm npm2.cmd
```

**Comment by I-keep-trying** [comment]
I just installed everything for the first time, so none of the above suggestions seem to apply to me.  I keep getting this when trying to execute `npx create-react-app`:
```
npx : The term 'npx' is not recognized as the name of a cmdlet, function, script file, or operable program. Check
the spelling of the name, or if a path was included, verify that the path is correct and try again.
```
Windows 10 Pro, build 1703, fresh re-install about 2 weeks ago with Developer Mode enabled
Node version 9.4.0
npm 5.6.0

The install instructions were a little confusing (there are multiple files to download, what is that no-install file for?) but I finally figured out that you have to turn nvm 'on' and then issue the 'use' command, (confused me because I'm not using a separate package manager so I thought that part didn't apply) but then I'm still getting the above error. 

I tried every possible thing. I uninstalled nvm and installed a different version, I checked the environment path, I ran out of ideas. I finally just did `npm install -g npx` and that seems to be working. Screw it.

**Comment by arfaWong** [comment]
I just remove the `npm` and `npm.cmd` files from the nodejs location, move dir `node_modules/npm` from the nodejs location to another location, and run `node npm-cli.js i -g npm@latest` inside `bin` dir in the moved folder.

**Comment by cannap** [comment]
@arfaWong  that worked for me! holy **** thank you

**Comment by pumano** [comment]
I got same problem when trying to *upgrade* my node **8.8.0** with npm **5.5.1** to node **9.8.0** but installation of node **9.8.0** does not update my npm from 5.5.1 to latest. After it I try to `npm i -g npm` and get error:

```
C:\Repository\myproject>npm i -g npm
npm WARN npm npm does not support Node.js v9.8.0
npm WARN npm You should probably upgrade to a newer version of node as we
npm WARN npm can't make any promises that npm will work with this version.
npm WARN npm Supported releases of Node.js are the latest release of 4, 6, 7, 8.
npm WARN npm You can find the latest version at https://nodejs.org/
WARNING: You are likely using a version of node-tar or npm that is incompatible with this version of Node.js.
Please use either the version of npm that is bundled with Node.js, or a version of npm (> 5.5.1 or < 5.4.0) or node-tar (> 4.0.1) that is compatible with Node.js 9 and above.
npm[19860]: src\node_zlib.cc:436: Assertion `args.Length() == 7 && "init(windowBits, level, memLevel, strategy, writeResult, writeCallback," " dictionary)"' failed.
 1: node::DecodeWrite
 2: node::DecodeWrite
 3: uv_loop_size
 4: v8::internal::wasm::SignatureMap::Find
 5: v8::internal::Builtins::CallableFor
 6: v8::internal::Builtins::CallableFor
 7: v8::internal::Builtins::CallableFor
 8: 000003BEBC8843C1
```

**Comment by ManifoldFR** [comment]
@arfaWong 's solution worked perfectly for me (I mixed a bit with @rolf-schmidiger 's solution and renamed instead of moving)

**Comment by rvkmar** [comment]
It is very simple.. If you are on windows kindly check https://github.com/felixrieseberg/npm-windows-upgrade
If you want to do it manually kindly check the documentation https://github.com/npm/npm/wiki/Troubleshooting#upgrading-on-windows

**Comment by noahleigh** [comment]
I threw together a PowerShell script to automate @arfaWong 's [solution](https://github.com/coreybutler/nvm-windows/issues/300#issuecomment-368192283): https://gist.github.com/noahleigh/ba34e18b3e0bc4a6a4e93ed7a480536e

Tested on PowerShell Core 6.0

**Comment by ZeVS777** [comment]
@rvkmar nope, no luck https://github.com/felixrieseberg/npm-windows-upgrade/issues/124

**Comment by ststeiger** [comment]
Same problem with
```
node-v8.6.0-win-x64
node-v9.9.0-win-x64
```
and if you rename the file or move them, it will fail with unable to create folder, please re-run as administrator. Appart from the fact that I don't have admin rights on this machine, it is doubtful that this would change anything, as the current user has mkdir permissions in that folder...

When will that ever get fixed ?

**Comment by ayvarot** [comment]
@arfaWong rocks!!

I used [cmder](http://cmder.net/), [nvm](https://github.com/coreybutler/nvm-windows)  1.1.6 and node 8.10.0
Updated from npm 5.6.0 to 5.8.0 :tada: 

`cd %programfiles%/nodejs`
`rm npm npm.cmd`
`mv node_modules/npm node_modules/npm2`
`node node_modules\npm2\bin\npm-cli.js i npm@latest -g`
`rm -rf npm2`

**Comment by sebtiz13** [comment]
thanks for your solution @ayvarot 
I used nvm 1.1.15 and node 9.8.0
Updated from 5.6.0 to 5.8.0

**Comment by mhoyer** [comment]
The solution by @ayvarot worked partially for me. Maybe due to the fact that I was running it on a msys git bash with ConEmu? Anyway, this is my adjusted snippet:
```
which npm && \
    cd `which npm | xargs dirname` && \
    mv npm npm2 && \
    mv npm.cmd npm2.cmd && \
    mv ./node_modules/npm ./node_modules/npm2 && \
    node node_modules/npm2/bin/npm-cli.js i npm@latest -g && \
    rm -r npm2 npm2.cmd ./node_modules/npm2

npm -v
```

**Comment by virtualvishwam** [comment]
@arfaWong  That worked...  Although not the ideal one...
Will wait for npm team to fix it...
.
Faced this with npm 5.6.0

**Comment by Jeff-Lewis** [comment]
A [plain Windows DOS version](https://gist.github.com/Jeff-Lewis/0714eabab700d3cbac2417c2ec28d0de) of @ayvarot's 

```shell
pushd %ProgramFiles%\nodejs
del npm npm.cmd
move node_modules\npm node_modules\npm2
node node_modules\npm2\bin\npm-cli.js i npm@latest -g
rd node_modules\npm2 /S /Q
popd
```

**Comment by lysla** [comment]
@arfaWong solution worked like a charm, thank you!

**Comment by stgaup** [comment]
As suggested:
`cd <your nodejs folder>
ren npm.cmd npm2.cmd
del npm
del npx

npm2 install npm -g

del npm2.cmd

npm -v`

**Comment by BurtHarris** [comment]
This seems now triggered by installation of Nodejs version 10.0.0 using nvm.

**Comment by RickieWoo** [comment]
@rolf-schmidiger worked for me! thank you

**Comment by kraihn** [comment]
I got around this by using Yarn, which I already had installed.

```
nvm use 10.0.0

# Have npm remove itself
npm uninstall -g npm

# Install npm into the yarn directory structure
yarn global add npm

# Use npm to install itself into the nodejs directory structure
npm install -g npm

# Remove the yarn installation of npm
yarn global remove npm
```

**Comment by hwangzhiming** [comment]
use yarn instead of npm it self, try `yarn global add npm@latest`, try to upgrade yarn if this fails.

**Comment by awarberg** [comment]
I also got the **Refusing to delete** error when trying to update npm using `npm install npm@latest --global`. 

The npm-windows-upgrade package was unable to update as well, suggesting to reinstall Node.js, in spite of this being a fresh unzip installation.

I was able to upgrade by moving the npm and npx files, as previously suggested, using this PowerShell script:

```
# Locate the Node.js installation
# containing the npm and npx files that are preventing the upgrade
Get-Command node.exe | % { Split-Path $_.Source } | cd

$CurrentNpmVersion = Get-ChildItem -Filter "npm.cmd" -File | % {
  Invoke-Expression "$($_.Name) --version"
}

# Version the current npm scripts to make room for the new version:
"npm", "npm.cmd", "npx", "npx.cmd" | %{ Get-ChildItem -Filter $_ } |
  Rename-Item -NewName { "$($_.BaseName)-$CurrentNpmVersion$($_.Extension)" }

# Use the current npm command to install the new version:
Invoke-Expression "npm-$($CurrentNpmVersion).cmd install --global npm@latest"
```

**Comment by lamuertepeluda** [comment]
Hi @awarberg I tried your script but I get this error. I tried with yarn as well, no luck... stupid Windows! 💢 

```
.\update_node.ps1
npm ERR! path C:\Program Files\nodejs\node_modules\npm\node_modules\worker-farm\node_modules\.bin\errno
npm ERR! code EEXIST
npm ERR! Refusing to delete C:\Program Files\nodejs\node_modules\npm\node_modules\worker-farm\node_modules\.bin\errno: is outside C:\Program Files\nodejs\node_modules\npm\node_modules\worker-farm\node_modules\errno and not a link
npm ERR! File exists: C:\Program Files\nodejs\node_modules\npm\node_modules\worker-farm\node_modules\.bin\errno
npm ERR! Move it away, and try again.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\vitom\AppData\Roaming\npm-cache\_logs\2018-05-11T12_55_55_902Z-debug.log
```

**Comment by awarberg** [comment]
@lamuertepeluda sorry to hear you still can't upgrade. The file mentioned `...\node_modules\.bin\errno` was not mentioned previously. This could be a corrupted installation due to multiple upgrade attempts. May be a good time to reinstall it, as npm-windows-upgrade suggests.

As for "stupid Windows" I don't agree. More likely, NPM was designed for Linux first, and Windows support was bolted on later, as an after thought. I see this regularly in other NPM packages as well. Typical problems are difficulty to install and requires manual and undocumented steps to ensure software prerequisites. Another common problem is case sensitive file system pathing.

**Comment by kufii** [comment]
I made a batch script to auto run @arfaWong's solution if anything interested.

```batch
@echo off

for /f %%i in ('npm -g -v') do set cur_version=%%i
for /f %%i in ('npm show npm version') do set latest_version=%%i

echo Updating to %latest_version%...

set node_path=%PROGRAMFILES%\nodejs
rename "%node_path%\npm" npm2
rename "%node_path%\npm.cmd" npm2.cmd
rename "%node_path%\node_modules\npm" npm2
node "%node_path%\node_modules\npm2\bin\npm-cli.js" i npm@latest -g

for /f %%i in ('npm -g -v') do set new_version=%%i

if "%new_version%" == "%latest_version%" (
	echo Successfully updated to %latest_version%. Cleaning up backups...
	del "%node_path%\npm2"
	del "%node_path%\npm2.cmd"
	@RD /S /Q "%node_path%\node_modules\npm2"
	echo Update complete.
) else (
	echo Something went wrong. Rolling back.
	if exist "%node_path%\npm" (
		del "%node_path%\npm"
	)
	if exist "%node_path%\npm.cmd" (
		del "%node_path%\npm.cmd"
	)
	if exist "%node_path%\node_modules\npm" (
		del "%node_path%\node_modules\npm"
	)
	rename "%node_path%\npm2" npm
	rename "%node_path%\npm2.cmd" npm.cmd
	rename "%node_path%\node_modules\npm2" npm
)

```

**Comment by johnmcase** [comment]
I created a gist from @kufii's bat script.  Slightly modified so that you can specify the version to install instead of always installing lastet.

https://gist.github.com/johnmcase/d31b799b9030327091a0e74880e4c530

**Comment by kufii** [comment]
@johnmcase good idea, I updated my comment so you can pass in the version you want to install as an argument (or "latest" to get the latest version)

**Comment by BurtHarris** [comment]
@kufii's updated version work's better for me than  having the desired number hardcoded into the script.

**Comment by SOSANA** [comment]
@kufii thank you!!

![notworthy](https://user-images.githubusercontent.com/2634185/40281672-b96fe7f8-5c32-11e8-9827-4a39fdbf6b42.gif)

**Comment by kamranayub** [comment]
@johnmcase Worked liked a charm, bookmarking.

**Comment by mkasprz** [comment]
Only one issue with this script found: when I selected to install `5.6` it successfully installed `5.6.0` and been rolling back as "5.6" differs from "5.6.0"...
Thank You for the script though!

**Comment by mazyvan** [comment]
Had the same issue again recently with nvm version 1.1.7
The solution of @arfaWong and @ayvarot worked the last time. But this time Idkw the npm folder wasn´t found. So I recommend not to delete the npm2 just in case.

**Comment by MendelBak** [comment]
I just spent nearly two days debugging errors caused by this problem and tried nearly every fix to upgrade my NPM from 5.6.0 to 6.1.0. @arfaWong's fix worked for me without any further issues. Thank you!

**Comment by apillai-incomm** [comment]
@arfaWong Works!!

**Comment by ctsstc** [comment]
I went about it the yarn method, but this should also work:

1. Grab npm from https://github.com/npm/cli/releases
1. Run the following from the directory you extracted it to
   - `node npm-cli.js i -g npm@latest`
   - It may complain about files needing to be deleted, for me I had my nvm and nodejs in a root directory on my drive so the standard program file location is inapplicable to me. I nuked the npm/npx/node_modules from my nodejs folder (I'm not afraid to redownload packages and I'm working with a fresh install now anyways).
1. It should install properly and you should be able to delete the folder you extracted and profit!

**Comment by DmitriyWebDev** [comment]
@ayvarot Thanks very much!!!!!!!! Your advice work for me on Windows 10.

**Comment by Fr0sT-Brutal** [comment]
Worked for me with just renaming `%Node%\npm.cmd`, `%Node%\npm` and `%Node%\npx.cmd`, **no directory renaming** was required.

**Comment by stephenasamoah** [comment]
@peternann's suggested solution here works, too.
https://github.com/coreybutler/nvm-windows/issues/389

For some reason the script above wasn't working for me and I was simply too tired resolving this to find out why.

**Comment by Fr0sT-Brutal** [comment]
This ridiculous bug still persists so I made a script that does the job. Must be placed inside Node folder where `npm.cmd` live.
```
:: Update NPM with workaround of "npm ERR! code EEXIST" bug https://github.com/coreybutler/nvm-windows/issues/300

@ECHO OFF

SETLOCAL

SET CDir=%~dp0%

REN "%CDir%\npm.cmd" "npm1.cmd"
DEL "%CDir%\npm"
DEL "%CDir%\npx.cmd"
DEL "%CDir%\npx"

CALL "%CDir%\npm1.cmd" update && DEL "%CDir%\npm1.cmd"
```

**Comment by hangvane** [comment]
when using the script of @kufii
Notice that the installation location should be replaced if you changed the path of node.js like me ,just replace the code  `set node_path=!PROGRAMFILES!\nodejs` to `set node_path=YOUR_NODE_PATH`

**Comment by kuncevic** [comment]
@kraihn [solution](https://github.com/coreybutler/nvm-windows/issues/300#issuecomment-386420934)does not work now with the latest yarn unless you add yarn bin in to the path:

1. get the bin path `yarn global bin`
2. stick that in to your path

**Comment by MrCroft** [comment]
Once a year I get excited, decide to install NVM on Windows thinking "this time it will work", but there's always something broken... Oh well, I guess it's back to the Node.js classic installer. I would use WSL, but unfortunately some Electron apps require you to have Node installed on the host OS.

Is this really unfixable by NVM alone?
As far as I've read in different threads, it would also be an npm or Node issue, but isn't there anything NVM alone could do about it? Otherwise, this issue kind of renders NVM useless on Windows.
Having to delete files ourselves while in the middle of the update process or even create scripts for that or whatever workarounds some people mentioned, is not a solution. It's hell!
The only thing I've tried (a comment in another thread, not this one, mentioning a few steps on how to use yarn to add npm) didn't work.

**Comment by terebentina** [comment]
@MrCroft while I agree that it would be nice to have it "just work", things are far from "hell" and even farther from being "useless".
I've been a happy user of this project for something like 2 years now.
This problem only occurs when you want to update the global npm itself. Installing new node versions via nvm has no problem and new nodes come with newer npm packages.
If you find yourself upgrading npm so often in your dev env that your life becomes hell because of this then you're probably doing something wrong.
I think only twice since I became a nvm user I had to actually upgrade npm itself and hit this problem. Not as bad as you make it sound.

**Comment by rsshilli** [comment]
Here's my script for upgrading npm on nvm: 

    https://stackoverflow.com/a/50955293/491553

I just did it on a new install.  Works great.

**Comment by ststeiger** [comment]
Maybe somebody should test the npm updater **WITHOUT ADMIN RIGHTS**...

**Comment by createdbyjurand** [comment]
cd C:\Users\niemczyj\AppData\Roaming\nvm\v6.17.0\node_modules\npm2\bin\
node npm-cli.js i -g npm@latest



I think this is more of a previous **npm** version bug than **nvm** one.

However, to update npm **delete** in `nvm/[node version]` folder (for me it is `v10.15.3`) following files:
```
C:\Users\[user name]\AppData\Roaming\nvm\[node version]\npm
C:\Users\[user name]\AppData\Roaming\nvm\[node version]\npm.cmd
C:\Users\[user name]\AppData\Roaming\nvm\[node version]\npx
C:\Users\[user name]\AppData\Roaming\nvm\[node version]\npx.cmd
```
**Rename** npm folder from
```
C:\Users\[user name]\AppData\Roaming\nvm\[node version]\node_modules\npm
```
to
```
C:\Users\[user name]\AppData\Roaming\nvm\[node version]\node_modules\npm2
```
Open any console with admin rights (or even right mouse click on vscode **shortcut**, than right mouse click on **Visual Studio Code** and hit **Run as administrator**) and type commands:
```
cd C:\Users\[user name]\AppData\Roaming\nvm\[node version]\node_modules\npm2\bin
node npm-cli.js i -g npm@latest
```
Done.

Delete `C:\Users\[user name]\AppData\Roaming\nvm\[node version]\node_modules\npm2` folder.

(This is just step by step reproduce what @19majkel94 said above)

**Comment by ststeiger** [comment]
@createdbyjurand: **WITHOUT** admin rights... 
You don't need admin rights to replace a bunch of file with another bunch of files...
(as long as you have write access to the files/directories)

**Comment by createdbyjurand** [comment]
> @createdbyjurand: **WITHOUT** admin rights...
> You don't need admin rights to replace a bunch of file with another bunch of files...
> (as long as you have write access to the files/directories)

I know. But, Program Files folder on Windows 10 has specific rights access. That is why lots of apps install itself lately in Roaming or Local folders. Even Visual Studio Code has two separate installs. One for Local and one for Program Files. One way to workaround this is to change in nvm nodejs folder location. Second, change Folder Security options but that makes this folder unsecure. What I wrote, what @19majkel94 wrote, is a solution to fix npm update just the way it is, using standard nvm install with default settings for people who just want to fix the problem without getting into more and more system or configuration changes. Because even if you change folder or access rights, you still have to do all the steps I listed, and the only change will be using the console without admin rights.

**Comment by ststeiger** [comment]
@createdbyjurand: Just saying, we run node in a corporate environment, and as developers, we have **no** admin rights. If an npm-update or a script therein requires admin rights, that breaks our dev-environment, and then we can't install any node packages, and then we can't work.

**Comment by Fr0sT-Brutal** [comment]
@createdbyjurand what do you need admin rights for?

**Comment by createdbyjurand** [comment]
> @createdbyjurand what do you need admin rights for?

Because without admin rights I get this:
```
$ npm i -g npm@latest
C:\Program Files\nodejs\npx -> C:\Program Files\nodejs\node_modules\npm\bin\npx-cli.js
npm ERR! path C:\Program Files\nodejs\npm
npm ERR! code EPERM
npm ERR! errno -4048
npm ERR! syscall open
npm ERR! Error: EPERM: operation not permitted, open 'C:\Program Files\nodejs\npm'
npm ERR!  { [Error: EPERM: operation not permitted, open 'C:\Program Files\nodejs\npm']
npm ERR!   cause:
npm ERR!    { Error: EPERM: operation not permitted, open 'C:\Program Files\nodejs\npm'
npm ERR!      errno: -4048,
npm ERR!      code: 'EPERM',
npm ERR!      syscall: 'open',
npm ERR!      path: 'C:\\Program Files\\nodejs\\npm' },
npm ERR!   stack:
npm ERR!    'Error: EPERM: operation not permitted, open \'C:\\Program Files\\nodejs\\npm\'',
npm ERR!   errno: -4048,
npm ERR!   code: 'EPERM',
npm ERR!   syscall: 'open',
npm ERR!   path: 'C:\\Program Files\\nodejs\\npm' }
npm ERR!
npm ERR! The operation was rejected by your operating system.
npm ERR! It's possible that the file was already in use (by a text editor or antivirus),
npm ERR! or that you lack permissions to access it.
npm ERR!
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator (though this is not recommended).

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\fehTH4RTH\AppData\Roaming\npm-cache\_logs\2019-03-22T14_28_26_153Z-debug.log
```

**Comment by Fr0sT-Brutal** [comment]
Oh I see. I never imagined having NPM installed in PF, for me is rather weird. I'd say those who do it are the reasons of their troubles :).

**Comment by createdbyjurand** [comment]
> Oh I see. I never imagined having NPM installed in PF, for me is rather weird. I'd say those who do it are the reasons of their troubles :).

Recently yes. Microsoft changed access rights for **Program Files** folder. That is why a lot of software is moving now towards **Roaming** or **Local** folder install directory. And this is why not so long ago everything was working fine with npm, and now suddenly we get this:
```
$ npm i -g npm@latest
C:\Program Files\nodejs\npx -> C:\Program Files\nodejs\node_modules\npm\bin\npx-cli.js
npm ERR! path C:\Program Files\nodejs\npm
npm ERR! code EPERM
npm ERR! errno -4048
npm ERR! syscall open
npm ERR! Error: EPERM: operation not permitted, open 'C:\Program Files\nodejs\npm'
npm ERR!  { [Error: EPERM: operation not permitted, open 'C:\Program Files\nodejs\npm']
npm ERR!   cause:
npm ERR!    { Error: EPERM: operation not permitted, open 'C:\Program Files\nodejs\npm'
npm ERR!      errno: -4048,
npm ERR!      code: 'EPERM',
npm ERR!      syscall: 'open',
npm ERR!      path: 'C:\\Program Files\\nodejs\\npm' },
npm ERR!   stack:
npm ERR!    'Error: EPERM: operation not permitted, open \'C:\\Program Files\\nodejs\\npm\'',
npm ERR!   errno: -4048,
npm ERR!   code: 'EPERM',
npm ERR!   syscall: 'open',
npm ERR!   path: 'C:\\Program Files\\nodejs\\npm' }
npm ERR!
npm ERR! The operation was rejected by your operating system.
npm ERR! It's possible that the file was already in use (by a text editor or antivirus),
npm ERR! or that you lack permissions to access it.
npm ERR!
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator (though this is not recommended).

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\fehTH4RTH\AppData\Roaming\npm-cache\_logs\2019-03-22T14_28_26_153Z-debug.log
```

**Comment by MrCroft** [comment]
@Fr0sT-Brutal well, for the sake of organization, I do it as well. If I already have a system folder designated for applications (Program Files), whenever I install anything - I install it in the "Program Files" folder. Especially not to pollute the drive's root folder with random folders like "nodejs" or "nvm" and also not to get to long paths, like "c:/Users/My Name/Roaming/ApplicationData/whatever", I just prefer "c:/Program Files/whatever".
Nonetheless, although it would have really bugged me to work like that (have nvm|node in a place outside program files), still... I've at least tried and I've installed it in c:/nvm (as for node versions, I've tried both c:/node and as a subfolder of nvm: c:/nvm/node something like that). But the issue persisted, npm still couldn't be updated by running `npm i -g npm`. Thank God, because I wouldn't want to live with things installed outside of the Program Files folder. Of course I finally gave up and still using 1 version of node at all times, installed using the default node installer for Windows from nodejs.org

**Comment by Fr0sT-Brutal** [comment]
@createdbyjurand recently? IIRC, PF became write-restricted for a user since XP.
@MrCroft I always wondered how people could live with all the stuff buried inside PF... Space in path means quoting is always required, madness with x64-PF and x32-PF adds one iteration for search, and - the most frustrating - a totally flat software structure. I've >100 apps even on home PC and +50 apps at work. I would jump out of a window if I had to search for apps inside PF.

**Comment by ststeiger** [comment]
You don't have to put it into 
`C:/Users/My Name/Roaming/Whatever`
you can use any directory as long as it is user-writeable. 

I have all programs on D:\Programs, none of them require admin rights, including nodejs, python, cmake, JetBrains, go, OpenOffice, etc.

Then I put them all into the path environment variables and registry for current_user.
Works like a charm. 
No crappy blackbox-installers required. 

If I need to find anything, I use locate32, I don't even browse the folders/Desktop/StartMenu.
Locate32 is just way faster. 
Or I run them via CTRL+R, as you can set the aliases in the registry in HKCU.

The only reason to have something in `C:\ProgramFiles` is when you want to install a certain application for every user on the machine, which as developer in most cases, you don't want, since having programs in systemwide path/registry might break the programs of other users, or inhibit their ability to use a different version than the version you use. 

By removing admin rights and keeping everything in user, you ensure that one user cannot break the the programs & settings of another user.  Also you ensure your applications run without admin rights, which is something that might come in handy at times. And at worst, you can always delete and purge a user account and recrate it later. 

That way I can run roughly 1'000 programs, all without a single minute spent installing. 
Even JetBrains IDE and PostgreSQL work that way. 
The only programs that I couldn't get to work like this are MS-Office, SQL-Server and Visual-Studio. 
npm shouldn't be added to this disreputable list, and all because it cannot be bothered to move a few existing files instead of overwriting them.

**Comment by chidveer** [comment]
I just renamed these files to "1"
npm1.cmd, npm1, npx1, npx1.cmd

and ran:
"**npm1 update -g**"

It looks like it worked fine, any thoughts?

**Comment by aaronmanela** [comment]
perhaps this can be built into nvm - perhaps by adding an npm version flag along with the node version and architecture flag on install, and an update-nvm function

**Comment by Brustulim** [comment]
> I just renamed these files to "1"
> npm1.cmd, npm1, npx1, npx1.cmd
> 
> and ran:
> "**npm1 update -g**"
> 
> It looks like it worked fine, any thoughts?

It worked for me too, but I had to delete the folder
\nvm\v10.15.3\node_modules\npm\node_modules\.bin
because the error pointing to this folder still appeared.

Thank you!

**Comment by dolpsdw** [comment]
Hey guys i have made a package for update the local npm version of Program Files\nodejs
Checkout
`npm install -g npm-win-upgrade`

**Comment by sodle** [comment]
@dolpsdw I get the following error when running `npm-win-upgrade`:

```
Npm instaled as Global package: uninstalling...                                                                     
(node:10192) UnhandledPromiseRejectionWarning: Error: Command failed: npm uninstall npm -g                          
npm ERR! path C:\Program Files\nodejs\npx.cmd                                                                       
npm ERR! code EEXIST                                                                                                
npm ERR! Refusing to delete C:\Program Files\nodejs\npx.cmd: is outside C:\Program Files\nodejs\node_modules\npm and
 not a link                                                                                                         
npm ERR! File exists: C:\Program Files\nodejs\npx.cmd                                                               
npm ERR! Move it away, and try again.                                                                               
                                                                                                                    
npm ERR! A complete log of this run can be found in:                                                                
npm ERR!     C:\Users\sjodle\AppData\Roaming\npm-cache\_logs\2019-04-23T17_42_11_305Z-debug.log                     
                                                                                                                    
    at ChildProcess.exithandler (child_process.js:281:12)                                                           
    at emitTwo (events.js:126:13)                                                                                   
    at ChildProcess.emit (events.js:214:7)                                                                          
    at maybeClose (internal/child_process.js:915:16)                                                                
    at Process.ChildProcess._handle.onexit (internal/child_process.js:209:5)                                        
(node:10192) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing
 inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (
rejection id: 1)                                                                                                    
(node:10192) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejec
tions that are not handled will terminate the Node.js process with a non-zero exit code.                            
```

**Comment by dolpsdw** [comment]
Hi @sodle,

The first error (node:10192 Command failed: npm uninstall npm -g): it happens because npm detected as global instaled in the %appdata%\npm folder, but this may be rest of and old node cache. This could be fixed with an npm list -g detection.

About the second error to me that error of 'Refusing to delete'... sounds like the Admin privileges where no authorized in the process. (UAC promp was denied?)

Try to execute the script and accept the Admin privilege scalation, and if that doesnt fix it please Tell me your node version of Origin to reproduce the bug.

**Comment by sodle** [comment]
@dolpsdw I'm pretty sure those are both part of the same error. At any rate, I no longer have the machine that the error occurred on. I'll let you know if I encounter the same issues on its replacement.

**Comment by Markus-Hanisch** [comment]
Hello @rsshilli 
thank you for providing the script. It works quite well and updates npm to the latest version as desired.
The only minor issue I faced, was that running `npm list -g --depth=0` to list all node packages resulted in two installations of npm.
So I simply added some lines of code to remove the "...-old" renamed files and the "...-old" renamed folder. 
Running `npm list -g --depth=0` to list all node packages resulted in one single installation of npm as desired.
Here is my updated version of your script:
`cd %APPDATA%\nvm\v8.9.0           # or whatever version you're using`
`mv npm npm-old`
`mv npm.cmd npm-old.cmd`
`cd node_modules\`
`mv npm npm-old`
`cd npm-old\bin`
`node npm-cli.js install -g npm@latest`

`cd %APPDATA%\nvm\v8.9.0           # or whatever version you're using`
`rm npm-old`
`rm npm-old.cmd`
`cd node_modules\`
`rm -rf npm-old`

> Here's my script for upgrading npm on nvm:
> 
> https://stackoverflow.com/a/50955293/491553
> 
> I just did it on a new install. Works great.

**Comment by OrangeDog** [comment]
@dolpsdw `npm-win-upgrade` failed with the same errors as `npm install -g npm`.
There is another package `npm-windows-upgrade`, which also fails, but doesn't say why (I assume it's the same).

**Comment by boenrobot** [comment]
@Markus-Hanisch @rsshilli Thank you for that script.

Taking that as a starting point, I'd like to add the following generalized powershell script, which allows you to upgrade NPM regardless of your current version:
```pwsh
$currentNvmVersion = ([regex]"^\s*\*\s*v?(\S+)").Match((nvm list | Out-String)).Groups[1].ToString()
Push-Location (Join-Path (Split-Path (Get-Command nvm).Path) ("v" + $currentNvmVersion))
  Move-Item npm npm-old
  Move-Item npm.cmd npm-old.cmd
  Move-Item npx.cmd npx-old.cmd
  Push-Location .\node_modules
    Move-Item npm npm-old
    Push-Location .\npm-old\bin
      node npm-cli.js install -g npm@latest
    Pop-Location
    Remove-Item npm-old -Recurse
  Pop-Location
  Remove-Item npx-old.cmd
  Remove-Item npm-old.cmd
  Remove-Item npm-old -Recurse
Pop-Location
```

Granted the fact is relies on the output of "nvm list" being in its current format, which makes it a little hacky, but still.

**Comment by jameslaydigital** [comment]
Read through this thread; in the end, I had the best luck uninstalling and reinstalling node.js using add/remove programs.

**Comment by kuncevic** [comment]
Problem was being fixed for me like a year ago by switching on mac ;)

**Comment by Fr0sT-Brutal** [comment]
@kuncevic 
> Problem was being fixed for me like a year ago by switching on mac ;)

Corrected:
Problem was being **avoided by me** like a year ago by switching on mac ;)
:rofl:

**Comment by ststeiger** [comment]
@kuncevic, @Fr0sT-Brutal: 
And by giving up your freedom of choice when it comes to hardware. 
Not a good trade-off at all. 
Also, you're giving up your control over the software that runs on your computer. 
If you need to switch somewhere, switch to Linux - that's the much better choice, for sever stuff anyway.

**Comment by darkship** [comment]
https://blog.npmjs.org/post/189618601100/binary-planting-with-the-npm-cli

I dropped nvm for windows. After 2 years, it still doesn't work properly even with the latest node version

**Comment by coreybutler** [comment]
Folks, this is an npm issue, not an NVM4W issue. There is a utility for helping with this, available at https://github.com/felixrieseberg/npm-windows-upgrade.

**Comment by xmedeko** [comment]
> I just renamed these files to "1"
> npm1.cmd, npm1, npx1, npx1.cmd
> 
> and ran:
> "**npm1 update -g**"

Just rename the file `npm1.cmd`, it's enough. Then run the `npm` update.

No need to use `npm-windows-upgrade`.

**Comment by bladeoflight16** [comment]
@coreybutler I'm not totally convinced of that. npm-windows-upgrade *did not* work for me, and after I manually cleared out the top level scripts (the same ones others are renaming here), `node (Join-Path $env:NVM_SYMLINK 'node_modules\npm\lib\npm.js') install -g npm` worked initially and `npm install -g npm` worked for a later update **without manual intervention**. It seems like the initial npm install is broken somehow.

Even if that is somehow npm's fault rather than nvm's, one step nvm could take to alleviate the problem is to clear out these scripts itself and invoke an update immediately when creating the initial installation. Or at least it could provide an option to do so.

**Comment by bladeoflight16** [comment]
@kuncevic Suggestion to improve your instructions:

Rather than copy files back and forth or try to move them around quickly enough while the process is running, just invoke npm's *script* directly with the node command.

PowerShell:

```
node (Join-Path $env:NVM_SYMLINK 'node_modules\npm\lib\npm.js') install -g npm
```

Command Prompt:

```
node %NVM_SYMLINK%\node_modules\npm\lib\npm.js install -g npm
```

**Comment by mattdamon108** [comment]
@bladeoflight16 I tried but got this error.

```cmd
C:\Users\woonk>node %NVM_SYMLINK%\node_modules\npm\lib\npm.js install -g npm
internal/modules/cjs/loader.js:796
    throw err;
    ^

Error: Cannot find module 'C:\Program'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:793:17)
    at Function.Module._load (internal/modules/cjs/loader.js:686:27)
    at Function.Module.runMain (internal/modules/cjs/loader.js:1043:10)
    at internal/main/run_main_module.js:17:11 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
```

**Comment by bladeoflight16** [comment]
@mattdamon108 Sorry. You need to quote the path in cmd if it has spaces. I updated the post. Thanks.

**Comment by anandshenoy14** [comment]
Was getting this Error on Windows and wanted to revert to an older npm for a project and in a very mild manner I just went to the already formed symlink location of npm command and renamed the npm folder in node_modules to its own version name and then downloaded the version I needed and created a folder called `npm` and dumped the downloaded contents there which worked!

**Comment by jakobrosenberg** [comment]
> > I just renamed these files to "1"
> > npm1.cmd, npm1, npx1, npx1.cmd
> > and ran:
> > "**npm1 update -g**"
> 
> Just rename the file `npm1.cmd`, it's enough. Then run the `npm` update.
> 
> No need to use `npm-windows-upgrade`.

@xmedeko @chidveer  Maybe I'm missing something here, but isn't ``npx npm-windows-upgrade`` more straight forward than manually renaming files before running ``npm update``?

**Comment by xmedeko** [comment]
@jakobrosenberg `npm-windows-upgrade` has not worked for me. Maybe I did something wrong, maybe i did [attempted upgrade](https://github.com/felixrieseberg/npm-windows-upgrade#fix-an-attempted-upgrade), or whatever. IMO one file rename is simpler than to study, why `npm-windows-upgrade` is not working and how to fix it.

Note: it's recommended to use `npm-windows-upgrade` globally, so you run it as `npm-windows-upgrade`, not `npx npm-windows-upgrade`.

**Comment by jakobrosenberg** [comment]
@xmedeko wouldn't ``npx npm-windows-upgrade`` run globally unless you're inside a node project with a local installation of ``npm-windows-upgrade``?

**Comment by xmedeko** [comment]
@jakobrosenberg Yes, you are right, I didn't know `npx` behave like that, I always use it for local project packages only.

**Comment by nokidding** [comment]
A batch script which works fine.
[https://gist.github.com/nokidding/aafaf90adc80cbce54b676340817bb13](https://gist.github.com/nokidding/aafaf90adc80cbce54b676340817bb13)

**Comment by ses403** [comment]
works like charm
just be sure 'set node_path=!PROGRAMFILES!\nodejs' is pointing correct path in your case
specifically incase of no standard installation(zip extract)

**Comment by ganselmo** [comment]
Same Issue! 
how i solved it:

1. Control Panel=>Uninstall a program
2. Delete NodeJS app
3. Delete All NodeJS Folder on program Files
4. Fresh Install Node JS

**Comment by Obiwarn** [comment]
I just needed Step 4, i.e. Just download and fresh install Node JS.
Then it worked.

**Comment by patricknelson** [comment]
Reinstalling the node version works fine, but since this is an issue associated with `nvm`, why not use the power of `nvm` to fix the problem? For me, instead of going through add/remove programs, it was way quicker/simpler to keep it all in the command line. For example:

```js
nvm uninstall VERSION
nvm install VERSION
```

After that, close and re-open the shell. Quick and easy (now that you know, at least). 😄 

**If you have any issues** (e.g. `nvm` indicates that the version is already installed), then check to see if the folder associated with that version still exists (e.g. `%NVM_HOME%\VERSION`) and that the file `node.exe` isn't still there. If it is still there, make sure you didn't already have any apps potentially running node (e.g. in my case my IDE was running it, so I closed that and was able to completely remove).

**Comment by sandyfigueroa** [comment]
This script works perfect: https://github.com/coreybutler/nvm-windows/issues/300#issuecomment-606682441
https://gist.github.com/nokidding/aafaf90adc80cbce54b676340817bb13

Thank you!

**Comment by liuliangsir** [comment]
This powershell script works perfect: [#300 (comment)](https://github.com/coreybutler/nvm-windows/issues/13#issuecomment-548933570)

Thanks!

**Comment by bladeoflight16** [comment]
I stopped using this tool in favor of [ps-nvm](https://github.com/aaronpowell/ps-nvm), but I had troubles with that as well. Almost all the errors I keep getting trying to upgrade npm are errors about files operations failing, probably because they're in use and thus locked by Windows. It's not like this doesn't make sense. npm is running, so of course its files are in use.

I don't like using non-standard solutions like random PowerShell scripts that refer to specific files npm installs. These are *highly* likely to break at some point as the package evolves.

A simple solution that has finally seemed to worked well for me is to stop using npm to upgrade itself. This obviously works since npm's files won't be in use if I use an entirely different package manager. I used yarn to upgrade it, and it worked fine. 

The one hiccup is that npm's version tagging is annoying: `latest` does not refer to the actual latest version, but to the latest *long term support* version. (This seems stupid to me, but whatever.) The same is true of its `next` tag. So you have to make sure you manually specify the latest version.


1. Install yarn:
    ```
    npm install -g yarn
    ```
2. Figure out the latest version of npm:
    ```
    yarn info npm dist-tags
    ```
    At the time of writing, the output looks like this:
    ```
    yarn info v1.22.10
    {
      latest: '6.14.11',
      next: '6.14.11',
      'latest-2': '2.15.12',
      'next-2': '2.15.12',
      'latest-1': '1.4.29',
      lts: '6.14.11',
      'latest-3': '3.10.10',
      'next-3': '3.10.10',
      'latest-4': '4.6.1',
      'next-4': '4.6.1',
      'latest-5': '5.10.0',
      'next-5': '5.10.0',
      'next-6': '6.14.11',
      'latest-6': '6.14.11',
      'next-7': '7.4.0'
    }
    ```
    You can see that the latest version available is 7.4.0, which is aliased as the `next-7` tag.
3. Finally, run the command to update npm, using the `@` syntax to specify the version you want:
    ```
    yarn global add npm@next-7
    ```

It's ridiculous that a package manager is designed in a manner that it can't update itself on one of the most popular operating systems, but this is the reality in which we live. As dumb as it sounds, maybe Node version managers on Windows should consider installing a second package manager by default.

**Comment by bladeoflight16** [comment]
@ganselmo If you're installing via the Control Panel, you're not using nvm. You need to *uninstall* that version before you even install nvm.

**Comment by bladeoflight16** [comment]
@jakobrosenberg 
> wouldn't `npx npm-windows-upgrade` run globally unless you're inside a node project with a local installation of `npm-windows-upgrade`?

This uses the npx executable script that you're trying to update, so it's likely to fail due to locking the file.

**Comment by begin-again** [comment]
I got this work by renaming `npm.cmd` to `npm1.cmd` and then deleted other npm files:  `npm`, `npx.cmd`, `npx`. Then I ran `npm1 install -g npm`

**Comment by MuhammadAbbasAkhtar** [comment]
> I had the same issue, and here is how I got through :
> 
> run `npm install -g npm` and **while it is running**:
> remove(backup) files `npm.cmd` and `npm` from `c:\Program Files\nodejs\`
> remove(backup) `.bin` folder from `c:\Program Files\nodejs\node_modules\npm\node_modules`, then it should succeed
> If you have `Refusing to delete`issues with any other npm module `files/folders` you have to just remove these and restart update process. To restart copy `npm.cmd` back to `c:\Program Files\nodejs\` . Then once `npm install -g npm` is running, remove `npm.cmd`, otherwise you get in to `Refusing to delete npm.cmd` issue, you might repeat that process until `npm install -g npm` is happy.
> 
> To remove the files you can use this shortcut: `mv npm.cmd "c:\Program Files\nodejs\node_modules\npm\node_modules"` as suggested by @farnetani
> 
> my env:
> `windows 7 x64, node 8.4, nvm 1.1.6`
> 
> ![image](https://user-images.githubusercontent.com/1283734/30268627-b417e194-9729-11e7-9c7e-aa103de70bbb.png)

Its been 3 years since this reply, but it still works! 
Thank you so much

**Comment by alekslario** [comment]
This was my fix:
1.Uninstall npm, node, remove all related directories.
2. Remove nvm bloat if any left
3. Install Volta
4. Everything works out of the box + build in rust
5. ???
6. Profit

**Comment by onx2** [comment]
> 
> 
> This was my fix:
> 1.Uninstall npm, node, remove all related directories.
> 2. Remove nvm bloat if any left
> 3. Install Volta
> 4. Everything works out of the box + build in rust
> 5. ???
> 6. Profit

Feels kind of mean to offer another package as a solution... but this absolutely worked out of the box for me and I love Rust. Thanks for the info @alekslario

**Comment by maximborisov** [comment]
I modified [this](https://stackoverflow.com/a/55091965) answer for node 15.10.0 and npm 7.5.3 and it very helpful for me. I added one another step for renaming npx too
```
cd %APPDATA%\nvm\v15.10.0           # or whatever version you're using
mv npm npm-old
mv npm.cmd npm-old.cmd
mv npx npx-old
mv npx.cmd npx-old.cmd
cd node_modules\
mv npm npm-old
cd npm-old\bin
node npm-cli.js i -g npm@latest
```

**Comment by median-dxz** [comment]
Another idea I've tried (I don't know if anyone has already mentioned it) is to install npm directly without using npm itself. Refer to [npmjs](https://www.npmjs.com/package/npm) in the “Direct Download” section, and it works.

**Comment by GorvGoyl** [comment]
> A batch script which works fine.
> https://gist.github.com/nokidding/aafaf90adc80cbce54b676340817bb13

This works. To use: 
1) download this as  `updateNpm.bat` file 
2) open powershell in that same folder and run this command ` ./updateNpm.bat latest`

**Comment by spaciandd** [comment]
> > A batch script which works fine.
> > https://gist.github.com/nokidding/aafaf90adc80cbce54b676340817bb13
> 
> This works. To use:
> 
> 1. download this as  `updateNpm.bat` file
> 2. open powershell in that same folder and run this command ` ./updateNpm.bat latest`

That worked for me, at last.

Thank you

**Comment by relu91** [comment]
`nvm` for Linux has this `install-latest-npm` command, maybe it is worth integrating the script above in `nvm-windows` as well.

**Comment by nsmithdev** [comment]
For me I only get the problem when updating npm from npm v6.
So using a newer version of npm via npx to run the upgrade works.

```
For the very newest version
npx npm install -g npm

Or use a specific version
npx npm@7 install -g npm@7
```

**Comment by SuperDzej** [comment]
Hi guys, have also problem to report after Windows reinstall can't seam to get latest npm for nvm node version 6.12.0 which I need to use, tried nvm version 1.1.9 and 1.1.7, it keeps failing with:

```
C:\Program Files\nodejs
`-- (empty)

npm ERR! Windows_NT 10.0.19042
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "install" "-g" "npm"
npm ERR! node v6.12.0
npm ERR! npm  v3.10.10
npm ERR! path C:\Program Files\nodejs\node_modules\.staging\@gar\promisify-f70bb42b
npm ERR! code ENOENT
npm ERR! errno -4058
npm ERR! syscall rename

npm ERR! enoent ENOENT: no such file or directory, rename 'C:\Program Files\nodejs\node_modules\.staging\@gar\promisify-f70bb42b' -> 'C:\Program Files\nodejs\node_modules\npm\node_modules\@gar\promisify'
npm ERR! enoent ENOENT: no such file or directory, rename 'C:\Program Files\nodejs\node_modules\.staging\@gar\promisify-f70bb42b' -> 'C:\Program Files\nodejs\node_modules\npm\node_modules\@gar\promisify'
npm ERR! enoent This is most likely not a problem with npm itself
npm ERR! enoent and is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! Please include the following file with any support request:
npm ERR!     C:\WINDOWS\system32\npm-debug.log
npm ERR! code 1
```

Tried all methods above all lead to this issue. Tried to make this folders, but it starts adding packages and then folder disappears and get same error.

**Comment by dschulten** [comment]
> This was my fix: 1.Uninstall npm, node, remove all related directories. 2. Remove nvm bloat if any left 3. Install Volta 4. Everything works out of the box + build in rust 5. ??? 6. Profit

Note that you need to enable Windows Developer Mode to allow Volta to set symlinks. Otherwise you get issues with access privileges similar to nvm. I have switched to Volta now but that might be helpful for nvm, too.

**Comment by devtobee** [comment]
Just delete the folder ...\AppData\Roaming\npm and \AppData\Roaming\npm-cache.
then try
npm install -g npm

**Comment by JimiC** [comment]
I was trying to upgrade `npm@6.14.16` which comes along with `node@14.19.1` and faced the same issue. 
I ended up using https://github.com/coreybutler/nvm-windows/issues/300#issuecomment-788810759 guide to make it finally upgrade.
Once it finished I noticed that the `node` folder now contained also some `ps1` scripts and any upgrade or downgrade of `npm` works fine after that. Probably because the system calls now the `ps1` script to start `npm` thus not getting blocked by the failure to delete the `cmd` files.

Using `node@16.13.1` with `npm@8.1.2` has no problem upgrading.

Looks like that from whatever version of `npm` the the team fixed this issue, `npm install -g npm@latest` just works.

**Comment by nagendragillalu** [comment]
Elevating user privileges to Administrator solved the issue for me.

**Comment by TrejoCode** [comment]
> 

### Usage:

```
.\updateNpm.bat <version>
```

Example:
```
.\updateNpm.bat 8.2.0
```

Thank  you 

> A batch script which works fine. https://gist.github.com/nokidding/aafaf90adc80cbce54b676340817bb13

**Comment by aczekajski** [comment]
How come this 5 years old issue is still not fixed..? I believe original nvm has a special command for upgrading npm and it should be the case here too. Having to rely on some obscure .bat scripts found in comments is ridiculous...

**Comment by Andrew-wong-ty** [comment]
- **I am a windows 10 user.**
-  I tried all the ways I can find, but they still do not work for me.  😥

And finally, I uninstall nodejs and then re-install it. Then everything is ok!  This is what I did:
1. Uninstall nodejs, and then delete the `environment variable path` like: `C:\Program Files\nodejs`
2. Install the latest version of nodejs from [here](https://nodejs.org/en/); ⚠️ you'd better not install nodejs in disk c.
3. After installing, type `node -v` and `npm -v` in `cmd`. If you can see the version, congratulations! No "npm ERR" anymore! If not, add the nodejs `environment variable path`.

**Comment by vinicioslc** [comment]
> 

This worked well for me ! volta.sh to the win!

**Comment by vinicioslc** [comment]
> This was my fix: 1.Uninstall npm, node, remove all related directories. 2. Remove nvm bloat if any left 3. Install Volta 4. Everything works out of the box + build in rust 5. ??? 6. Profit

This worked well for me ! volta.sh to the win!

**Comment by coreybutler** [comment]
Yes folks, volta is an option, as is nodist, fnm, nvs, etc. You're welcome to use those. I'm going to remove comments about other version managers though, because they don't help this project improve... and that's the point of the issues for this project.

For anyone finding this issue from Google, please make sure you've read the instructions... specifically the ones about removing any prior versions of Node.js (https://github.com/coreybutler/nvm-windows/wiki/Common-Issues#uninstall-existing-node-installation-before-installing-nvm4w). It is possible to update npm, but you cannot have conflicts in your PATH, or permission restrictions... and there are a few older versions of npm with hardcoded pathing issues. When the v1.1.11 patch release drops, use the `debug` function to check your environment for potential problems.

I still cannot replicate this problem and there has been a `help wanted` label for 5 years. I believe it when people say they are still running into npm update problems, but without an environment to test on, it's really hard to troubleshoot. At this point though, with active development proceeding on [Runtime](https://github.com/coreybutler/nvm-windows/wiki/Runtime) and inconsistency in failures, it's time to close this issue. If anyone runs into this again, please open a new issue and provide details. It's important to know things like the specific build of Windows, the locale, whether it's running in an Active Directory controlled environment or not, whether developer mode is on/off, which shell is running the command, whether the shell is run as an administrator or not, whether there are special characters in your install path or not, whether Node was installed without a version manager before or not, whether another version manager was used before NVM4W, etc.

**Comment by MaxiSantos** [comment]
I asked CHATGPT and it suggested to upgrade npm with `npm install -g npm@latest`

If you're using nvm-windows to manage your Node.js versions on Windows and you want to update npm to X version, you can follow these steps:

1. Open a Command Prompt or PowerShell window.

2. Make sure you're using the Node.js version you want to update by running the following command:
   ```
   nvm use <version>
   ```
   Replace `<version>` with the Node.js version you want to use. 
3. To update npm, you can use the following command:
   ```
   npm install -g npm@X-version
   ```

   This will install the X version of npm globally on your system.

4. Once the installation is complete, verify the npm version by running:
   ```
   npm --version
   ```

   You should see the X version of npm that you installed.

By following these steps, you should be able to update npm to the X version using nvm-windows on Windows.

Follow this url to know which version of npm belongs to node
https://nodejs.org/en/download/releases
